### PR TITLE
Allow floats in commands

### LINKED ISSUE #77: Floats are intentionally not supported
Referring to the comment here: https://github.com/Nordix/eredis/blob/7d9f67fe9e11b3a696b9e10e55ec7fcac99d4973/src/eredis.erl#L230 - it seems like an oddly specific thing to block floats from being added as parameters to `q` and `qp`.

I'm using Redis TimeSeries, and floats are absolutely necessary here. Converting my terms to strings has been an easy bypass, but the behavior strikes me as an incorrect and possibly outdated opinion, that has no place in a general purpose library.

Thanks for providing this library - currently testing it for performance improvements over my previous goto.

### PR DESCRIPTION
**Pull request by dsschneidermann**
Closes #77

### CODE REVIEW

**Comment by dsschneidermann** [comment]
I did not update the CHANGELOG.md as that seems to be done at-release time.

**Comment by dsschneidermann** [comment]
Whoops, almost got the test wrong, force-pushed now.

**Comment by dsschneidermann** [comment]
I was convinced that `[short]` was the correct option as the default was giving me scientific notation. But reading the doc page on it a little further:
```When the float is inside the range (-2⁵³, 2⁵³), the notation that yields the smallest number of characters is used (scientific notation or normal decimal notation). Floats outside the range (-2⁵³, 2⁵³) are always formatted using scientific notation to avoid confusing results when doing arithmetic operations.```

So what is the correct option to pass to `float_to_binary` that will enforce a decimal encoding? Should it be `[{decimals,100},compact]`? Advice is welcome, I haven't pushed Erlang code before.

**Comment by zuiderkwast** [comment]
I see your comment now. :) I don't think any 64 bit float needs more than 20 decimals so i dont think we need 100.

If we use 'short' we need conditional code for otp 25. I'm fine with that, also fine without it. :)

There is much to read about floats... If you don't have a good suggestion, i think float_to_binary/1 is fine and possibly use 'short' for otp >= 25.

**Comment by dsschneidermann** [comment]
Haha, yeah, so much for this being super easy.

AFAIK, Redis fails here when it encounters a float that cannot be represented as a 64-bit float, at least for Redis TimeSeries (I presume it is the same all across the product). Here is an example doc: https://redis.io/commands/ts.add

> The double number should follow [RFC 7159](https://tools.ietf.org/html/rfc7159) (JSON standard). In particular, the parser rejects overly large values that do not fit in binary64. It does not accept NaN or infinite values.

In particular, scientific notation is not allowed. The linked RFC states:
```
Since software that implements
   IEEE 754-2008 binary64 (double precision) numbers [[IEEE754](https://www.rfc-editor.org/rfc/rfc7159#ref-IEEE754)] is
   generally available and widely used, good interoperability can be
   achieved by implementations that expect no more precision or range
   than these provide, in the sense that implementations will
   approximate JSON numbers within the expected precision.  A JSON
   number such as 1E400 or 3.141592653589793238462643383279 may indicate
   potential interoperability problems, since it suggests that the
   software that created it expects receiving software to have greater
   capabilities for numeric magnitude and precision than is widely
   available.

   Note that when such software is used, numbers that are integers and
   are in the range [-(2**53)+1, (2**53)-1] are interoperable in the
   sense that implementations will agree exactly on their numeric
   values.
```

**Comment by dsschneidermann** [comment]
@zuiderkwast How do you feel about `list_to_binary(io_lib_format:fwrite_g(X))`?

Looking at various packages that have to perform the same work, https://github.com/michalmuskala/jason/blob/master/CHANGELOG.md - a popular JSON library for Elixir.

> Use the :erlang.float_to_binary(_, [:short]) function, instead of io_lib_format.fwrite_g/1 where available (OTP 24.1+). This provides equivalent output with much less memory used and significantly improved performance.

Also fwrite_g is used by Elixir to stringify floats by default. I'm not exactly up for adding compilation targets and testing if the short format is available (the JSON library mentioned runs a test for it on startup with a try/catch).

**Comment by dsschneidermann** [comment]
Looking at the Elixir roadmap, switching to use the `short` option is planned once the last supported version reaches OTP-25. So there is consensus here that using the `short` options is the right longterm choice.

I've added a bunch of edge case tests and have also just tested with redis `TS.ADD` that it accepts a large float (greater than 2^53), which is encoded by both the given approaches as '1.801439851e16'. I'm happy with the code and tests that are now force-pushed, only using the `fwrite_g` approach, but the OTP-25 specific support would be nice-to-have.

File: test/eredis_tests.erl [hunk 1]
```diff
@@ -10,10 +10,14 @@ multibulk_test_() ->
...
                    list_to_binary(create_multibulk(['SET', foo, 123]))),
 
-     ?_assertThrow({cannot_store_floats, 123.5},
-                   list_to_binary(create_multibulk(['SET', foo, 123.5])))
-    ].
+     %% Test floats by float_to_binary(Float, [short]) or io_lib_format:fwrite_g(Float)
+     ?_assertEqual(<<"*3\r\n$3\r\nSET\r\n$3\r\nfoo\r\n$5\r\n123.5\r\n">>,
+                   list_to_binary(create_multibulk(['SET', foo, 123.5]))),
+     ?_assertEqual(<<"*3\r\n$3\r\nSET\r\n$3\r\nfoo\r\n$17\r\n3.141592653589793\r\n">>,
+                   list_to_binary(create_multibulk(['SET', foo, 3.141592653589793238462643383279]))),
```
Comment (zuiderkwast) [inline test/eredis_tests.erl:20]: Redis just stores these as strings. To make redis store a float internally without the external module (time series), use a sorted sets, ZADD and ZSCORE. The *score* is stored as a double.

Example: ZADD k 3.141592653589793238462643383279 foo. Debug example using telnet:

```
$ telnet 127.0.0.1 6379
Trying 127.0.0.1...
Connected to 127.0.0.1.
Escape character is '^]'.
ZADD k 3.141592653589793238462643383279 foo
:1
ZSCORE k foo
$17
3.141592653589793
```

Btw, it seems (from my manual testing) to accept all the formats we have come up with, so I guess `io_lib:format("~.20g", Float)` is fine, for example.

We could add some roundtrip tests, starting with a float in Erlang, store in Redis, get it back to Erlang and check that we get the same float. But then we face another odd problem: Erlang requires a float to contain a dot, so it doesn't accept "4e100" and "4" but requires "4.0e100" and "4.0". It's not really a problem in eredis right now (since we don't support RESP3) but I guess you're using floats in you application and you want to get the values back to floats in Erlang, don't you? (I don't know what's the best approach for that. Try-catch with a fallback to inserting ".0" at the right place maybe...)

**Review by zuiderkwast — COMMENTED** [review]
> AFAIK, Redis fails here when it encounters a float that cannot be represented as a 64-bit float

That shouldn't be possible to get from Erlang since Erlang's float is a double (AKA binary 64-bit float).

> In particular, scientific notation is not allowed.

That sounds weird. JSON surely allows numbers like 4e100 which is in the part of the RFC you quoted. That is scientific notation, isn't it?

Where does it say Redis doesn't accept scientific notaion? Or what am I missing?

File: src/eredis.erl [hunk 2]
```diff
@@ -226,13 +226,11 @@ create_multibulk(Args) ->
...
-%% as we do not want floats to be stored in Redis. Your future self
-%% will thank you for this.
+%% @doc: Convert given value to binary. Fallbacks to term_to_binary/1.
 to_binary(X) when is_list(X)    -> list_to_binary(X);
 to_binary(X) when is_atom(X)    -> atom_to_binary(X, utf8);
 to_binary(X) when is_binary(X)  -> X;
 to_binary(X) when is_integer(X) -> integer_to_binary(X);
-to_binary(X) when is_float(X)   -> throw({cannot_store_floats, X});
+to_binary(X) when is_float(X)   -> list_to_binary(io_lib_format:fwrite_g(X));
+%% OTP-25: to_binary(X) when is_float(X)   -> float_to_binary(X, [short]);
```
[the remark this answers is not in the document]
Comment (dsschneidermann) [reply src/eredis.erl:235]: That's correct, Elixir releases are pinned to support a specific OTP version.

[the remark this answers is not in the document]
Comment (dsschneidermann) [reply src/eredis.erl:235]: Oh and yes, you're right that scientific notation is fine, that was my mistake and I noted it in a later comment :)

File: test/eredis_tests.erl (hunk 1 above)
Comment (dsschneidermann) [reply test/eredis_tests.erl:20]: That's a great catch. I'll try to update this PR with the alterations for ZADD and ZSCORE.

It's indeed a problem if Redis can return unparseable formats.. The large numbers aren't so relevant for my use case with timeseries metrics here, but I'll have to test it and see if Elixir adds anything to the parsing..

File: src/eredis.erl (hunk 2 above)
[the remark this answers is not in the document]
Comment (dsschneidermann) [reply src/eredis.erl:235]: Updated to use `io_lib:format` with the "~.20g" format string ✔️ 

Adding an OTP_RELEASE condition will also need the same branching during tests, so it can match the new "short" format in OTP 25. I'll think I can add it if you want, but it's your burden to maintain over time :) Using only `io_lib:format` is good enough for me, and it's predictably 20 decimal digits.

File: test/eredis_tests.erl (hunk 1 above)
Comment (dsschneidermann) [reply test/eredis_tests.erl]: Elixir handles this carefully with an `add_dot` function, https://github.com/elixir-lang/elixir/blob/3d98c56e69df86db3861f2035ede71d0cd83aa5b/lib/elixir/lib/float.ex#L182

```
:erlang.binary_to_float(add_dot(acc, dot?))
```

It's definitely still a huge issue for user-code.. and further adds support to the idea that throwing an error from `eredis` might be suitable, if only to notify users of this potential one-in-a-million issue :P If RESP3 support is in the long-term plan for this library, it could be a solution in the interim. Otherwise, we add a note in the README?

RE: ZADD and ZSCORE, as right now there seems to be no infrastructure in place for round-trip tests to Redis, I couldn't easily add this :) The only tests for this PR remain the ones exercising the `create_multibulk` function.

File: src/eredis.erl (hunk 2 above)
[the remark this answers is not in the document]
Comment (zuiderkwast) [reply src/eredis.erl]: In the tests, can't we just compare the result as a float? There's no need to check the textual representation, is there?

File: test/eredis_tests.erl [hunk 3]
```diff
@@ -10,10 +10,14 @@ multibulk_test_() ->
...
-
      ?_assertEqual(<<"*3\r\n$3\r\nSET\r\n$3\r\nfoo\r\n$3\r\n123\r\n">>,
                    list_to_binary(create_multibulk(['SET', foo, 123]))),
 
-     ?_assertThrow({cannot_store_floats, 123.5},
-                   list_to_binary(create_multibulk(['SET', foo, 123.5])))
-    ].
+     %% Test floats formatting with io_lib:format.
+     ?_assertEqual(<<"*3\r\n$3\r\nSET\r\n$3\r\nfoo\r\n$21\r\n123.50000000000000000\r\n">>,
+                   list_to_binary(create_multibulk(['SET', foo, 123.5]))),
```
Comment (zuiderkwast) [inline test/eredis_tests.erl:18]: Always 20 digits? It means even a small number like 0.0 or 1.0 needs 21 bytes. Seems like a waste, but floats are not everywhere so perhaps it's acceptable. I'd prefer shorter version like "123.5" though...

File: src/eredis.erl (hunk 2 above)
[the remark this answers is not in the document]
Comment (zuiderkwast) [reply src/eredis.erl]: I see tests are already comparing RESP protocol representation, so I guess we should keep it like that. Well, I guess we can live with an OTP_RELEASE condition also in the tests.

Another idea: How about using `io_lib_format:fwrite_g(X)` for OTP < 25 and `short` for OTP >= 25? It will not break in future versions. :-)

File: test/eredis_tests.erl (hunk 3 above)
Comment (dsschneidermann) [reply test/eredis_tests.erl:18]: Totally agree on that and it makes sense that OTP-25 adds a new standard solution. I think it's a problem if the library provides different behavior when OTP-25 is used.

Would it be acceptable to use the internal `fwrite` call when OTP-25 is not available, and OTP-25 always when it is? That way, we don't need a conditional for the tests as they produce the same result - and never something unexpected, like postfix zeroes.

Comment (zuiderkwast) [reply test/eredis_tests.erl:18]: :+1: Exactly this I just wrote in another comment. We only use `io_lib_format:fwrite_g/1` in old OTP versions, there is no risk it will break in the future.

Comment (dsschneidermann) [reply test/eredis_tests.erl]: 🏁 Done, should work for all of OTP < 21 and current + future versions

**Comment by dsschneidermann** [comment]
And the last fix is to remove the binaries for `elvis` and `rebar3` that I was keeping in this branch, for some reason >:)

**Comment by zuiderkwast** [comment]
The fmt CI job complains about indentation. I guess you're not using emacs. :)

File: test/eredis_tests.erl [hunk 4]
```diff
@@ -10,10 +10,14 @@ multibulk_test_() ->
...
-                   list_to_binary(create_multibulk(['SET', foo, 123.5])))
-    ].
+     %% Test floats by float_to_binary(Float, [short]) or io_lib_format:fwrite_g(Float)
+     ?_assertEqual(<<"*3\r\n$3\r\nSET\r\n$3\r\nfoo\r\n$5\r\n123.5\r\n">>,
+                   list_to_binary(create_multibulk(['SET', foo, 123.5]))),
+     ?_assertEqual(<<"*3\r\n$3\r\nSET\r\n$3\r\nfoo\r\n$17\r\n3.141592653589793\r\n">>,
+                   list_to_binary(create_multibulk(['SET', foo, 3.141592653589793238462643383279]))),
+     ?_assertEqual(<<"*3\r\n$3\r\nSET\r\n$3\r\nfoo\r\n$14\r\n1.801439851e16\r\n">>,
+                   list_to_binary(create_multibulk(['SET', foo, 18014398510000000.0])))
+   ].
```
Comment (zuiderkwast) [inline test/eredis_tests.erl:23]:
```suggestion
    ].
```

File: src/eredis.erl [hunk 5]
```diff
@@ -226,13 +226,21 @@ create_multibulk(Args) ->
...
+%% @doc: Convert a float to binary shortest representation.
+-ifdef(OTP_RELEASE).
+  -if(?OTP_RELEASE >= 25).
+    encode_float(X) -> float_to_binary(X, [short]).
+  -else.
+    encode_float(X) -> list_to_binary(io_lib_format:fwrite_g(X)).
+  -endif.
+-else.
+    encode_float(X) -> list_to_binary(io_lib_format:fwrite_g(X)).
+-endif.
```
Comment (zuiderkwast) [inline src/eredis.erl:246]:
```suggestion
%% @doc: Convert a float to binary shortest representation.
-ifdef(OTP_RELEASE).
-if(?OTP_RELEASE >= 25).
encode_float(X) -> float_to_binary(X, [short]).
-else.
encode_float(X) -> list_to_binary(io_lib_format:fwrite_g(X)).
-endif.
-else.
encode_float(X) -> list_to_binary(io_lib_format:fwrite_g(X)).
-endif.
```

**Comment by dsschneidermann** [comment]
> The fmt CI job complains about indentation. I guess you're not using emacs. :)

Nope, it's been a learning experience with Erlang in vscode. Thanks!

**Comment by alexandremcosta** [comment]
This is great! Thanks a lot for the work!
